### ISSUE TITLE
Add "Domain Validation Method" indicator to certificate profiles

### ISSUE THREAD

**Issue by clintwilson** [cabforum/servercert#459]
(I believe this has been discussed in the past, but I wasn't able to find a related tracking issue. I'm not proposing this as a new priority, just writing it down while I happened to be thinking about it.)

### Problem Statement
It would be beneficial to Subscribers and Relying Parties to be able to identify the method used to validate the domain names present in a certificate. 

### Pros/Cons
These data would provide evidence of compliance by the CA, allow Subscribers to verify processes by which their domains are being validated meet expectations, and enable Relying Parties to better understand how certificates are being validated/issued across the ecosystem.
Conversely, the extra data may not be worth the bits necessary to represent them, certificates with large numbers of SANs validated using multiple methods may prove particularly difficult to design around, and Subscribers may not want this information publicly disclosed.

### Strawdog Proposal(s)
When presented in the appropriate context, the domain validation method can be adequately and accurately represented by a single number matching the subsection of BR 3.2.2.4 which defines the domain validation method used.
If there's support, but not consensus, for introducing this practice, it could be added as an option (instead of requirement) within the BRs so there's alignment/consistency on presentment.

- The method number could be appended to the BR reserved policy OID already included in the certificate (i.e. instead of the current OID, it would be this new OID to represent both certificate validation type and domain validation method)
  - This would increase certificate size by a very small amount
  - This may limit inclusion to only a single domain validation method
  - The full set of possible OIDs would still need to be enumerated and reserved, which would be a lot of OIDs across the 4 certificate validation types currently reserved
  - Interaction with ASSs would need to be verified to ensure certificates correctly parse with the certificate validation type enforcement/presentment
- The method number(s) could be included as new policy OIDs under the CA/BF arc reserved for this purpose
  - This would take up more bits
  - This would allow for multiple domain validation methods to be represented
- The method number(s) could be added as a new extension
- The method number(s) could be added, using any of the above mechanisms, only to precertificates

**Comment by vanbroup** [comment]
> It would be beneficial to Subscribers and Relying Parties to be able to identify the method used to validate the domain names present in a certificate.

Subscribers should know how their domain names were verified as they are involved, why would they want this in the certificate, do we have any data supporting this request?

I'm also not sure if it's adding any substantial value to relying parties, some argue that subject information or policy qualifiers are not understood or looked at by relying parties, can you try to explain why would this be valuable and understood by relying parties?

On the other hand, I can imagine that this information can be interesting to the ecosystem, assuming it would support a clear purpose or goal, but doubt that the certificate is the right place to source this statistical information, especially where we used to trim every bit for efficiency.

**Comment by barrini** [comment]
TBD at the next F2F

**Comment by birgelee** [comment]
Dear CA/Browser Forum community,

I understand @vanbroup 's point that including this information in the certificate increases certificate size which is an important metric to consider. Additionally, I also respect the point that doing this without a clear reason/motivation does not make sense as it introduces additional complexity/data for no reason.

That said, I did want to chime in with a research/threat intelligence perspective to say that to many members of the community this would be incredibly valuable, and I believe it would have a concrete positive impact on catching miss issuance and cyber attacks.

From the threat intelligence perspective: having reviewed several attacks that involved miss issued certs and worked with members of the threat intelligence and attack response community, knowledge about how a certificate was validated is often one of the most important parts of an investigation into a fraudulently-issued certificate. Not only can validation method allow investigators to identify trends across attacks and better understand adversarial behavior, but it can often help identify the root cause of the miss issuance. For example, say a certificate was validated by DNS change, this immediately rules out several attack vectors like compromises at the victim's webserver. Similarly, we are now aware of a major vulnerability introduced by stale whois data which could potentially lead to miss issuance. Knowing a cert was validated via whois or not single-handedly can point to or point away from this vulnerability being the root cause. I know CAs do collect this information on the back end and potentially subpoenas can be used to gather this information when law enforcement is involved. However, I urge everyone to think about the practicality of these channels and their ability to hinder an active investigation. Not only does the effectiveness of subpoenas significantly reduce when there is a multi-national element to an investigation (subpoenas need to be issued in the country that has jurisdiction over the CA which may not actually be involved in the other elements of the investigation), but the delay and headache of needing to go through law enforcement effectively makes this information inaccessible in a timely manner to an investigation. As I am sure everyone realizes the ability to respond rapidly to an ongoing investigation is crucial and having this information in a certificate (or even precertificate) is extremely helpful.

From a research perspective: we are constantly analyzing and monitoring the way the PKI operates. CT has been a huge enabler for PKI research. However, CTs limitations at exploring validation are quite notable. From the CT perspective, there is simply a stream of certificates issued but no information about how they have been validated. For several past projects (and an ongoing project) we have relied on our partnership with Let's Encrypt to share validation log data from which we publish appropriate summary statistics from with approval (see https://www.usenix.org/conference/usenixsecurity23/presentation/cimaszewski and https://www.usenix.org/conference/usenixsecurity21/presentation/birge-lee ). This data has substantially more information than what is being proposed to be put in a cert, but even the minimal amount of information that could fit into a cert or precert (e.g., validation BR section number) would be extremely helpful and could enable us to scale much of our analysis beyond Let's Encrypt to the entire PKI ecosystem. CT log data is becoming a very popular data source for PKI research, and I am sure other research groups as well could do very valuable analysis if validation information were in CT. To give a concrete example, there is some discussion on MDSP about Whois-based validation and the extent to which it can be securely relied on. If the CA/B Forum were to take any action regarding Whois-based, validation a core statistic to inform this decision would be the number of certs validated using this method. This may be collected (rather inaccurately) via CPS scraping and inquiries to CAs about how often they use the method if they use it. If validation data was in CT, we could have an exact number on this statistic in seconds.

I have wanted validation information in the certificate or in CT for a long time now, and I am happy it is finally being discussed. I do not think this info needs to be in the cert itself as I think it is unlikely a browser will make an accept/reject decision based on this info (although in extreme cases like when TLS-SNI was found vulnerable, a validation method could be effectively disabled instantly with a browser update using this info). However, I do think it is extremely helpful for it to be logged in a public manner. I also do not entirely buy the certificate size argument because 1) the number of bytes needed is minimal compared to the size of the post-quantum safe asymmetric keys that need to be stored in certs and 2) this info could be put only in precerts eliminating the need to inflate the final TLS cert at all. Having the method in the precert only would largely allow for both the treat intel and research usages I outlined above.

TL;DR Having this info would let us better catch the perpetrators of cyber attacks and would greatly improve our ability to  research the PKI which will improve community decisions in the long run.

Best,
Henry Birge-Lee
Research Software Engineer, Princeton University

**Comment by CBonnell** [comment]
It is not possible to encode ancillary information in a precertificate without including it in the final certificate. See RFC 6962, section 3.2 "tbs_certificate" on the allowed differences between a precertificate and final certificate.

Absent some new mechanism to omit fields in a final certificate that are present in a final certificate, this validation information will need to be present in final certificates as well.

**Comment by aaomidi** [comment]
> as I think it is unlikely a browser will make an accept/reject decision based on this info 

I wouldn't make this assumption personally. I'm of the opinion that certificates shouldn't include any data that is not using for the relying party to drive trust decisions. 

It would be reasonable for a "safer" browser to drive trust decisions based off the validation method. Even if the major browsers never inspect this field, some other browsers can (and probably will) make trust decisions based off the validation method.

I wrote a bit more about my thoughts on this here: https://webpki.substack.com/p/issuance-auditability-certificate

And agreed, even if it was possible to put different values in precert vs cert - it would significantly increase the complexity of precerts vs certs.

**Comment by birgelee** [comment]
Thanks @CBonnell for clarifying.  I had noticed this proposal in @clintwilson 's original bullet list, but I agree given the limitations you discussed it doesn't entirely make sense.

**Comment by orangepizza** [comment]
Would bitmask format for this make sense? Put entire variation methods as single bitstring, and if cab forum method N is ised 2^Nth digit is set.
It'd reduce certificate size increase into 10~20bytes, and using multiple methods in single certificate doesn't increase cert size.
Because numbering we'd still need an Oid per SAN types, but multi type SAN certificates are rare and unless it's dnsname and ipaddress it wouldn't make sense

**Comment by birgelee** [comment]
@orangepizza I generally like the bitmask idea you are proposing. I was just curious about two things:

1. You are proposing a single bitmask per SAN type not per SAN, so if one domain in a cert is DNS validated and one domain is HTTP validated, both the HTTP and DNS bit would be set. We would not be able to differentiate which domain was validated which way but we would know both validation methods were used. In general I think this should be sufficient, but I am curious as to how much larger things would actually be if the validation method for each SAN were enumerated. Would this not be roughly 1 byte a SAN? is a lot more space needed for structure? 
2. You are proposing a bitstring per SAN type. Is this preferable to simply having a longer bitstring that includes all methods for all SAN types? I guess its theoretically smaller because you only need the bitstring for the SAN types being used. However, I wonder if the formatting to indicate which bitstring is being used will add more bytes.

**Comment by orangepizza** [comment]
@birgelee 
1. it would be a byte per SAN, but only if we use implicit encoded array of bytes that assumed ordered same as SAN list: otherwise we have to write entire SANs again in this extension.
2. I thought about it but it's possible to method number increasing (like it happens on http-v2 and tls-alpn) and new SAN type pops up sometime (while not in public space) I couldn't think of encoding that wouldn't be per type anyway.

**Comment by aaomidi** [comment]
IMO mixed validations shouldn't be in the same certificate.

**Comment by orangepizza** [comment]
@aaomidi I think mixed validation in a same certificate can't be avoided in some cases, due to other types of SAN there exist:
for example entire point of cert for (dv) onion domain is linking it to surfacenet entity, but they are validated by methods under appendix B, not something under 3.2.2.4:, those just can't have same validation method. same thing applies to raw IP certificate:

**Comment by aaomidi** [comment]
That's an interesting property to take advantage of. FWIW I do think that may be considered a misuse of what the certificate is supposed to do in this case.

SANs being included in the certificate don't really mean they're "associated" with each other than the subscriber had control over many of these at the same time. That doesn't mean that these SANs share the same "stewardship" if that makes sense?

For example, free load balancers sometimes put multiple unrelated SANs in the same certificate because it's cheaper to batch them together than to get a new certificate per SAN. This doesn't mean that these domains should be assumed to at all be related to each other.

**Comment by birgelee** [comment]
I think @orangepizza 's proposal of one byte per SAN is good.

Regardless if it is deemed strategic to limit mixed-validation certificates or not, I do not particularly think it makes sense to introduce in this context. A ballot to add a one-byte field for validation method essentially just requires CAs to add a new certificate extension and then encode data into it that they are already obligated to log. Nothing customer facing needs to be changed.

Eliminating mixed-validation certificates prohibits a behavior CAs may currently be doing and the change is more substantial and customer facing. I would personally not intertwine these two proposals even if it does reduce 1-byte per SAN.

BTW: there is a concrete methodology for encoding onion domains into certificates which does exploit the concept of SANs in a certificate having some association: https://www.sauteed-onions.org/
I feel there is some precedent for this type of use.

**Comment by daknob** [comment]
Hello everyone,

I'd like to add some quick comments on this proposal idea overall:

1) The more data on the entire domain (organization / extended / ...) validation process becomes live, the better it is for the ecosystem to effectively audit its operations.

2) The more we minimize the cost of these additions, the easier it is to adopt them.

With that said, I'd like them to be as complete as possible, as too many iterations on certificate profiles may increase the costs over time to beyond acceptable for the value brought.

The method used for domain validation is useful, but it'd be interesting to know when the validation occurred. Maybe the CA is reusing a 1-year old validation for the first two domains, a 7 month validation for the third one, and it just recently validated the last 5 of them in the last hour.

As reuse can exist for these very large periods of time, if this is isn't encoded somehow, I feel like the value of this field will be less than originally intended.

If CA X made a mistake between August 2nd and August 22nd in 3.2.2.4.7, we'd have no way of knowing purely based on this new addition and the `notBefore` timestamp (with its own well known issues) if a certificate has been issued in violation of the policy or not.

Threat Intelligence usecases remain, and security too, but we'd have to take into account this problem above. Maybe the attacker gained access to DNS or the web server once, and now it's really patched, but they can just reuse their validation for a very extended period of time, and you're out there trying to figure out how come they still have access to the services.

**Comment by clintwilson** [comment]
For reference, this proposal was last presented within a draft ballot in 2018 which, after discussion at the time, came to the conclusion to use 2 extensions (one for DNS Names and one for IP Addresses), each with defined bits for the Validation Methods as presented by their respective subsection numbers under Sections 3.2.2.4 and 3.2.2.5 of the TBRs.

At that time, as I understand it, both for certificate size considerations and overall perceived value of the varying possible degrees of nuance that could be represented, the proposal was to just include all methods used for any SAN in the extension. Presumably, this would be just 1 method for a vast majority of certificates, but certainly for some large, shared SAN certificates, this could be many of the methods allowed by the TBRs at a given point.

Solutioning here can, of course, be approached in a lot of ways (as the above discussion demonstrates), but it currently seems to me that a similar approach as proposed in 2018 would still provide the best cost-to-value ratio. The incremental value of having a breakdown of the method used on a per-SAN basis or the timestamp of each validation's completion, while absolutely notable, do seem to me to capture most of their value in different types of edge cases (or nearly-edge cases, at least). Relevant here, I think, are efforts to reduce the maximum allowed validity of certificates along with paring down the total set of allowed validation methods based on both security principles and ability to support necessary modern automation. What are folks' views on that opinion statement? 

One other point that I think is relevant to this discussion is that introducing these extensions to certificate profiles would also necessitate being much more careful about making high-impact changes to extant methods, and would encourage (strongly) a practice of introducing new methods rather than altering those already in the TBRs.

For some additional reading, here are a few relevant messages (imo), but the surrounding email threads are also full of discussion and additional context:
- https://archive.cabforum.org/pipermail/validation/2018-August/001062.html
- https://archive.cabforum.org/pipermail/validation/2018-August/000990.html
- https://archive.cabforum.org/pipermail/validation/2018-October/001086.html

**Comment by orangepizza** [comment]
If we use bitmask we need to assign some bits for onion address validation methods. Because they are DNS SAN but it doesn't govern by 3.2.2.4, but method from appendix B: while most are point back to 3.2.2.4, CSR signed by onion key doesn't.
as use later number may collide if we add new method in 3.2.2.4, I think using a number from retired one is better idea:
I think bit 5, 3.2.2.4.5 (Domain Authorization Document) can used for CSR one as loosely what that method does

**Comment by wthayer** [comment]
> What are folks' views on that opinion statement?
> it currently seems to me that a similar approach as proposed in 2018 would still provide the best cost-to-value ratio.
I agree.

**Comment by romanf** [comment]
(Full disclosure: Imwork for a CA but the following is my private opinion.)

Given the facts that certificate lifetime ist going down and probably only DNS and ACME based validation will survive, I personally don't think it's worth spending all the time and energy in developing, implementing, bug-fixing, ... this in the long run.

-Roman

**Comment by aaomidi** [comment]
I don't see how that makes this less useful tbh.

**Comment by dzacharo** [comment]
It is very important that contributions submitted to the CA/B Forum are protected by the IPR Policy. I suggest all individuals that want to contribute to improving the CA/B Forum TLS Guidelines [join](https://cabforum.org/about/information/potential-members/) the Server Certificate Working Group, as [Interested Parties](https://cabforum.org/about/membership/interested-parties/).

I noticed some contributions in this GitHub issue from non-Members that will not be taken into consideration without a signed IPR agreement from the contributors.
